File: tools/string_analyzer.py
#!/usr/bin/env python3

"""
String Analyzer Tool - Python Implementation
Extracts ASCII and Unicode strings from binaries using the `strings` command
Analyzes string content for security-relevant patterns
"""

import json
import sys
import os
import argparse
import subprocess
import re
from pathlib import Path
from typing import Dict, List, Any

class StringAnalyzer:
    def __init__(self, binary_path: str, min_length: int = 4,
                 encoding: str = 'all', search_pattern: str = ''):
        self.binary_path = binary_path
        self.min_length = min_length
        self.encoding = encoding
        self.search_pattern = search_pattern

    def extract_strings(self) -> Dict[str, Any]:
        """Extract strings using the Linux `strings` command"""
        try:
            if not Path(self.binary_path).exists():
                raise FileNotFoundError(f"Binary not found: {self.binary_path}")

            cmd = ['strings']

            # Encoding flags
            if self.encoding == 'ascii':
                cmd.append('-a')
            elif self.encoding == 'unicode':
                cmd.extend(['-e', 'l'])
            elif self.encoding == 'utf16':
                cmd.extend(['-e', 'L'])
            elif self.encoding == 'all':
                cmd.extend(['-a'])

            # Minimum length
            if self.min_length:
                cmd.extend(['-n', str(self.min_length)])

            cmd.append(self.binary_path)

            result = subprocess.run(
                cmd,
                capture_output=True,
                text=True,
                timeout=30
            )

            if result.returncode != 0:
                return {
                    'error': result.stderr.strip(),
                    'summary': f'strings command failed: {result.stderr.strip()}'
                }

            all_strings = [s.strip() for s in result.stdout.split('\n') if s.strip()]

            # Filter by search pattern if provided
            if self.search_pattern:
                try:
                    pattern = re.compile(self.search_pattern, re.IGNORECASE)
                    filtered = [s for s in all_strings if pattern.search(s)]
                except re.error:
                    filtered = [s for s in all_strings if self.search_pattern.lower() in s.lower()]
            else:
                filtered = all_strings

            return self._analyze_strings(filtered, all_strings)

        except Exception as e:
            return {'error': str(e), 'summary': f'Failed to extract strings: {str(e)}'}

    def _analyze_strings(self, strings: List[str], all_strings: List[str]) -> Dict[str, Any]:
        """Analyze extracted strings for security-relevant patterns"""
        # Security-relevant string categories
        security_patterns = {
            'buffer_overflow': [r'(?i)(buffer|overflow|stack|heap)'],
            'bounds_checking': [r'(?i)(bounds|limit|length|size|count)'],
            'error_messages': [r'(?i)(error|fail|invalid|abort|denied)'],
            'validation': [r'(?i)(validate|check|verify|sanitize|allow)'],
            'format_strings': [r'%[dsxun]'],
            'function_names': [r'(?i)(strcpy|strncpy|sprintf|snprintf|memcpy|memmove|gets|fgets)'],
            'paths': [r'^/(etc|usr|var|tmp|dev|proc|sys)/'],
            'ip_addresses': [r'\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}'],
            'urls': [r'https?://'],
        }

        categories: Dict[str, List[str]] = {}
        for category, patterns in security_patterns.items():
            matches = []
            for s in strings:
                for pattern in patterns:
                    if re.search(pattern, s):
                        matches.append(s)
                        break
            if matches:
                categories[category] = matches[:20]

        # Statistics
        total = len(all_strings)
        filtered_count = len(strings)
        unique = len(set(all_strings))

        # Longest strings
        sorted_strings = sorted(strings, key=len, reverse=True)
        longest_strings = sorted_strings[:10] if sorted_strings else []

        return {
            'total_strings': total,
            'unique_strings': unique,
            'filtered_count': filtered_count if self.search_pattern else total,
            'min_length': self.min_length,
            'encoding': self.encoding,
            'search_pattern': self.search_pattern or None,
            'categories': categories,
            'longest_strings': longest_strings,
            'sample_strings': sorted_strings[:100] if len(sorted_strings) <= 100 else sorted_strings[:100],
            'summary': f'Found {total} strings ({unique} unique) in {self.binary_path}'
        }


def main():
    parser = argparse.ArgumentParser(description='String Analyzer Tool')
    parser.add_argument('--binary', required=True, help='Path to binary file')
    parser.add_argument('--min-length', type=int, default=4, help='Minimum string length')
    parser.add_argument('--encoding', choices=['all', 'ascii', 'unicode', 'utf16'], default='all',
                        help='String encoding')
    parser.add_argument('--search-pattern', help='Regex pattern to filter strings')
    parser.add_argument('--output-format', choices=['json', 'text'], default='json')

    args = parser.parse_args()

    try:
        analyzer = StringAnalyzer(
            binary_path=args.binary,
            min_length=args.min_length,
            encoding=args.encoding,
            search_pattern=args.search_pattern or ''
        )

        analysis_result = analyzer.extract_strings()

        output = {
            'data': analysis_result,
            'summary': analysis_result.get('summary', 'String analysis completed'),
            'metadata': {
                'binary': args.binary,
                'min_length': args.min_length,
                'encoding': args.encoding,
                'search_pattern': args.search_pattern,
                'categories_found': list(analysis_result.get('categories', {}).keys()),
                'ai_next_step': 'Compare strings between before/after versions to identify security-relevant changes'
            }
        }

        if args.output_format == 'json':
            print(json.dumps(output, indent=2))
        else:
            print(f"String Analyzer Report")
            print(f"=====================")
            print(f"Binary: {args.binary}")
            print(f"Total Strings: {analysis_result.get('total_strings', 0)}")
            print(f"Unique Strings: {analysis_result.get('unique_strings', 0)}")
            print(f"\nCategories Found:")
            for cat, strs in analysis_result.get('categories', {}).items():
                print(f"  {cat}: {len(strs)} occurrences")

    except Exception as e:
        error_output = {
            'error': str(e),
            'summary': f'String analysis failed: {str(e)}',
            'metadata': {'ai_next_step': 'Check binary file and strings command availability'}
        }
        print(json.dumps(error_output, indent=2))
        sys.exit(1)


if __name__ == '__main__':
    main()

File: tools/string_analyzer.ts
#!/usr/bin/env node

/**
 * String Analyzer Tool - TypeScript Wrapper (Refactored for OpenCode)
 * Extracts and analyzes strings from binaries via FastAPI server.
 */

interface StringAnalyzerConfig {
  binary_path: string;
  min_length?: number;
  encoding?: 'all' | 'ascii' | 'unicode' | 'utf16';
  search_pattern?: string;
}

interface ToolResult {
  status: 'success' | 'error';
  tool_name: string;
  execution_time_ms: number;
  data: any;
  summary: string;
  metadata: { ai_next_step?: string; [key: string]: any };
}

async function runStringAnalyzer(config: StringAnalyzerConfig): Promise<ToolResult> {
  const startTime = Date.now();

  try {
    if (!config.binary_path) {
      throw new Error('binary_path is required');
    }

    const payload: Record<string, any> = { binary: config.binary_path };
    if (config.min_length !== undefined) payload.min_length = config.min_length;
    if (config.encoding) payload.encoding = config.encoding;
    if (config.search_pattern) payload.search_pattern = config.search_pattern;

    const apiUrl = 'http://tools:8000/string_analyzer';
    let response;
    try {
      response = await fetch(apiUrl, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(payload),
        signal: AbortSignal.timeout(60000)
      });
    } catch (fetchError) {
      const errorMsg = fetchError instanceof Error ? fetchError.message : String(fetchError);
      return {
        status: 'error', tool_name: 'string_analyzer',
        execution_time_ms: Date.now() - startTime,
        data: { error: errorMsg },
        summary: `Failed to connect to FastAPI server at ${apiUrl}: ${errorMsg}`,
        metadata: { ai_next_step: 'Ensure the tools container is running on http://tools:8000' }
      };
    }

    if (!response.ok) {
      let errorDetail = '';
      try {
        const errorBody = await response.json();
        errorDetail = errorBody.detail || JSON.stringify(errorBody);
      } catch {
        errorDetail = await response.text();
      }
      return {
        status: 'error', tool_name: 'string_analyzer',
        execution_time_ms: Date.now() - startTime,
        data: { error: `HTTP ${response.status}`, details: errorDetail },
        summary: `FastAPI server returned error: ${response.status} - ${errorDetail}`,
        metadata: { ai_next_step: 'Verify binary file exists and is readable on tools container' }
      };
    }

    let responseData;
    try {
      responseData = await response.json();
    } catch {
      const rawText = await response.text();
      return {
        status: 'error', tool_name: 'string_analyzer',
        execution_time_ms: Date.now() - startTime,
        data: { raw: rawText },
        summary: 'Failed to parse FastAPI response as JSON',
        metadata: { ai_next_step: 'Check server logs for string extraction errors' }
      };
    }

    return {
      status: 'success',
      tool_name: 'string_analyzer',
      execution_time_ms: Date.now() - startTime,
      data: responseData,
      summary: `Successfully extracted strings from ${config.binary_path} (${responseData.count || '?'} strings found)`,
      metadata: {
        ai_next_step: 'Review strings for security-relevant content (error messages, URLs, version info, credentials)'
      }
    };
  } catch (error) {
    const message = error instanceof Error ? error.message : String(error);
    return {
      status: 'error', tool_name: 'string_analyzer',
      execution_time_ms: Date.now() - startTime,
      data: { error: message },
      summary: `String analyzer error: ${message}`,
      metadata: { ai_next_step: 'Check configuration and ensure binary file is valid' }
    };
  }
}

async function main() {
  const args = process.argv.slice(2);
  const config: StringAnalyzerConfig = { binary_path: '' };

  for (let i = 0; i < args.length; i++) {
    switch (args[i]) {
      case '--binary': config.binary_path = args[++i]; break;
      case '--min-length': config.min_length = parseInt(args[++i], 10); break;
      case '--encoding': config.encoding = args[++i] as StringAnalyzerConfig['encoding']; break;
      case '--search-pattern': config.search_pattern = args[++i]; break;
    }
  }

  const result = await runStringAnalyzer(config);
  console.log(JSON.stringify(result, null, 2));
}

if (require.main === module) {
  main().catch((err) => {
    console.error(JSON.stringify({
      status: 'error', tool_name: 'string_analyzer', execution_time_ms: 0,
      data: { error: err.message },
      summary: `Fatal error: ${err.message}`,
      metadata: { ai_next_step: 'Check logs' }
    }, null, 2));
    process.exit(1);
  });
}

export { runStringAnalyzer, StringAnalyzerConfig, ToolResult };